python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1381 @@
+    subset of columns to materialize.
+    """
+
+    cdef:
+        shared_ptr[CScannerBuilder] wrapped
+        CScannerBuilder* builder
+
+    def __init__(self, Dataset dataset not None, ScanContext context not None):
+        cdef shared_ptr[CScannerBuilder] builder
+        builder = make_shared[CScannerBuilder](
+            dataset.unwrap(), context.unwrap()
+        )
+        self.init(builder)
+
+    cdef void init(self, shared_ptr[CScannerBuilder]& sp):
+        self.wrapped = sp
+        self.builder = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CScannerBuilder]& sp):
+        cdef ScannerBuilder self = ScannerBuilder.__new__(ScannerBuilder)
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CScannerBuilder] unwrap(self):
+        return self.wrapped
+
+    def project(self, columns):
+        """Set the subset of columns to materialize.
+
+        This subset will be passed down to DataSources and corresponding
+        DataFragments. The goal is to avoid loading, copying, and deserializing
+        columns that will not be required further down the compute chain.
+
+        Raises exception if any column name does not exists in the dataset's
+        Schema.
+
+        Parameters
+        ----------
+        columns : list of str
+            List of columns to project. Order and duplicates will be preserved.
+
+        Returns
+        -------
+        self : ScannerBuilder
+        """
+        cdef vector[c_string] cols = [tobytes(c) for c in columns]
+        check_status(self.builder.Project(cols))
+        return self
+
+    def finish(self):
+        """Return the constructed now-immutable Scanner object"""
+        return Scanner.wrap(GetResultValue(self.builder.Finish()))
+
+    def filter(self, Expression filter_expression not None):
+        """Set the filter expression to return only rows matching the filter.
+
+        The predicate will be passed down to DataSources and corresponding
+        DataFragments to exploit predicate pushdown if possible using partition
+        information or DataFragment internal metadata, e.g. Parquet statistics.
+
+        Raises exception if any of the referenced columns does not exists in
+        the dataset's schema.
+
+        Parameters
+        ----------
+        filter_expression : Expression
+            Boolean expression to filter rows with.
+
+        Returns
+        -------
+        self : ScannerBuilder
+        """
+        check_status(self.builder.Filter(filter_expression.unwrap()))
+        return self
+
+    def use_threads(self, bint value):
+        """Set whether the Scanner should make use of the thread pool.
+
+        Parameters
+        ----------
+        value : boolean
+
+        Returns
+        -------
+        self : ScannerBuilder
+        """
+        check_status(self.builder.UseThreads(value))
+        return self
+
+    def schema(self):
```

Comment:
Should this be a property? (or is it something that needs to be calculated (potentially expensive))

Reply:
Sure, property SGTM

Reply:
Done.